File: android/src/main/kotlin/com/ahmtydn/contacts_bridge/ContactsBridgePlugin+Handlers.kt
package com.ahmtydn.contacts_bridge

import android.content.ContentResolver
import android.content.ContentUris
import android.database.Cursor
import android.net.Uri
import android.provider.ContactsContract
import android.provider.ContactsContract.CommonDataKinds.*
import android.provider.ContactsContract.Data
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.*

fun ContactsBridgePlugin.handleGetAllContacts(call: MethodCall, result: Result) {
    val args = call.arguments as? Map<String, Any> ?: mapOf()
    val withProperties = args["withProperties"] as? Boolean ?: false
    val withThumbnail = args["withThumbnail"] as? Boolean ?: false
    val withPhoto = args["withPhoto"] as? Boolean ?: false
    val sorted = args["sorted"] as? Boolean ?: true

    coroutineScope.launch(Dispatchers.IO) {
        try {
            val contacts = getAllContacts(withProperties, withThumbnail, withPhoto, sorted)
            withContext(Dispatchers.Main) {
                result.success(contacts)
            }
        } catch (e: Exception) {
            withContext(Dispatchers.Main) {
                result.error("FETCH_ERROR", "Failed to fetch contacts", e.message)
            }
        }
    }
}

fun ContactsBridgePlugin.handleGetContact(call: MethodCall, result: Result) {
    val args = call.arguments as? Map<String, Any>
    val contactId = args?.get("id") as? String

    if (contactId == null) {
        result.error("INVALID_ARGUMENTS", "Contact ID is required", null)
        return
    }

    val withProperties = args["withProperties"] as? Boolean ?: true
    val withThumbnail = args["withThumbnail"] as? Boolean ?: false
    val withPhoto = args["withPhoto"] as? Boolean ?: false

    coroutineScope.launch(Dispatchers.IO) {
        try {
            val contact = getContact(contactId, withProperties, withThumbnail, withPhoto)
            withContext(Dispatchers.Main) {
                result.success(contact)
            }
        } catch (e: Exception) {
            withContext(Dispatchers.Main) {
                result.error("FETCH_ERROR", "Failed to fetch contact", e.message)
            }
        }
    }
}

fun ContactsBridgePlugin.handleSearchContacts(call: MethodCall, result: Result) {
    val args = call.arguments as? Map<String, Any>
    val query = args?.get("query") as? String

    if (query == null) {
        result.error("INVALID_ARGUMENTS", "Search query is required", null)
        return
    }

    val withProperties = args["withProperties"] as? Boolean ?: false
    val sorted = args["sorted"] as? Boolean ?: true

    coroutineScope.launch(Dispatchers.IO) {
        try {
            val contacts = searchContacts(query, withProperties, sorted)
            withContext(Dispatchers.Main) {
                result.success(contacts)
            }
        } catch (e: Exception) {
            withContext(Dispatchers.Main) {
                result.error("SEARCH_ERROR", "Failed to search contacts", e.message)
            }
        }
    }
}

fun ContactsBridgePlugin.handleCreateContact(call: MethodCall, result: Result) {
    val args = call.arguments as? Map<String, Any>
    val contactData = args?.get("contact") as? Map<String, Any>

    if (contactData == null) {
        result.error("INVALID_ARGUMENTS", "Contact data is required", null)
        return
    }

    coroutineScope.launch(Dispatchers.IO) {
        try {
            val contact = createContact(contactData)
            withContext(Dispatchers.Main) {
                result.success(contact)
            }
        } catch (e: Exception) {
            withContext(Dispatchers.Main) {
                result.error("CREATE_ERROR", "Failed to create contact", e.message)
            }
        }
    }
}

fun ContactsBridgePlugin.handleUpdateContact(call: MethodCall, result: Result) {
    val args = call.arguments as? Map<String, Any>
    val contactData = args?.get("contact") as? Map<String, Any>

    if (contactData == null) {
        result.error("INVALID_ARGUMENTS", "Contact data is required", null)
        return
    }

    coroutineScope.launch(Dispatchers.IO) {
        try {
            val contact = updateContact(contactData)
            withContext(Dispatchers.Main) {
                result.success(contact)
            }
        } catch (e: Exception) {
            withContext(Dispatchers.Main) {
                result.error("UPDATE_ERROR", "Failed to update contact", e.message)
            }
        }
    }
}

fun ContactsBridgePlugin.handleDeleteContact(call: MethodCall, result: Result) {
    val args = call.arguments as? Map<String, Any>
    val contactId = args?.get("id") as? String

    if (contactId == null) {
        result.error("INVALID_ARGUMENTS", "Contact ID is required", null)
        return
    }

    coroutineScope.launch(Dispatchers.IO) {
        try {
            deleteContact(contactId)
            withContext(Dispatchers.Main) {
                result.success(null)
            }
        } catch (e: Exception) {
            withContext(Dispatchers.Main) {
                result.error("DELETE_ERROR", "Failed to delete contact", e.message)
            }
        }
    }
}

File: darwin/Classes/ContactsBridgePlugin+Handlers.swift
#if os(macOS)
    import Cocoa
    import FlutterMacOS
#endif
#if os(iOS)
    import Flutter
    import UIKit
#endif
import Contacts
import ContactsUI

@available(iOS 9.0, macOS 10.11, *)
extension ContactsBridgePlugin {
    
    // MARK: - Permission Management
    
    func handleRequestPermission(call: FlutterMethodCall, result: @escaping FlutterResult) {
        DispatchQueue.global(qos: .userInteractive).async {
            CNContactStore().requestAccess(for: .contacts) { (granted, error) in
                DispatchQueue.main.async {
                    if let error = error {
                        result(FlutterError(
                            code: "PERMISSION_ERROR",
                            message: "Failed to request permission",
                            details: error.localizedDescription
                        ))
                        return
                    }
                    result(granted ? "granted" : "denied")
                }
            }
        }
    }
    
    func handleGetPermissionStatus(result: @escaping FlutterResult) {
        let status = CNContactStore.authorizationStatus(for: .contacts)
        switch status {
        case .notDetermined:
            result("notDetermined")
        case .restricted:
            result("restricted")
        case .denied:
            result("denied")
        case .authorized:
            result("authorized")
        case .limited:
            result("limited")
        @unknown default:
            result("unknown")
        }
    }
    
    // MARK: - Contact Operations
    
    func handleGetAllContacts(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments", details: nil))
            return
        }
        
        let withProperties = args["withProperties"] as? Bool ?? false
        let withThumbnail = args["withThumbnail"] as? Bool ?? false
        let withPhoto = args["withPhoto"] as? Bool ?? false
        let sorted = args["sorted"] as? Bool ?? true
        
        DispatchQueue.global(qos: .userInteractive).async {
            do {
                let contacts = try self.getAllContacts(
                    withProperties: withProperties,
                    withThumbnail: withThumbnail,
                    withPhoto: withPhoto,
                    sorted: sorted
                )
                let contactsData = contacts.map { self.contactToDict($0) }
                DispatchQueue.main.async {
                    result(contactsData)
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "FETCH_ERROR",
                        message: "Failed to fetch contacts",
                        details: error.localizedDescription
                    ))
                }
            }
        }
    }
    
    func handleGetContact(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let contactId = args["id"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Contact ID is required", details: nil))
            return
        }
        
        let withProperties = args["withProperties"] as? Bool ?? true
        let withThumbnail = args["withThumbnail"] as? Bool ?? false
        let withPhoto = args["withPhoto"] as? Bool ?? false
        
        DispatchQueue.global(qos: .userInteractive).async {
            do {
                if let contact = try self.getContact(
                    id: contactId,
                    withProperties: withProperties,
                    withThumbnail: withThumbnail,
                    withPhoto: withPhoto
                ) {
                    let contactData = self.contactToDict(contact)
                    DispatchQueue.main.async {
                        result(contactData)
                    }
                } else {
                    DispatchQueue.main.async {
                        result(nil)
                    }
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "FETCH_ERROR",
                        message: "Failed to fetch contact",
                        details: error.localizedDescription
                    ))
                }
            }
        }
    }
    
    func handleSearchContacts(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let query = args["query"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Search query is required", details: nil))
            return
        }
        
        let withProperties = args["withProperties"] as? Bool ?? false
        let sorted = args["sorted"] as? Bool ?? true
        
        DispatchQueue.global(qos: .userInteractive).async {
            do {
                let contacts = try self.searchContacts(
                    query: query,
                    withProperties: withProperties,
                    sorted: sorted
                )
                let contactsData = contacts.map { self.contactToDict($0) }
                DispatchQueue.main.async {
                    result(contactsData)
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "SEARCH_ERROR",
                        message: "Failed to search contacts",
                        details: error.localizedDescription
                    ))
                }
            }
        }
    }
    
    func handleCreateContact(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let contactData = args["contact"] as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Contact data is required", details: nil))
            return
        }
        
        DispatchQueue.global(qos: .userInteractive).async {
            do {
                let contact = try self.createContact(from: contactData)
                let createdContactData = self.contactToDict(contact)
                DispatchQueue.main.async {
                    result(createdContactData)
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "CREATE_ERROR",
                        message: "Failed to create contact",
                        details: error.localizedDescription
                    ))
                }
            }
        }
    }
    
    func handleUpdateContact(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let contactData = args["contact"] as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Contact data is required", details: nil))
            return
        }
        
        DispatchQueue.global(qos: .userInteractive).async {
            do {
                let contact = try self.updateContact(from: contactData)
                let updatedContactData = self.contactToDict(contact)
                DispatchQueue.main.async {
                    result(updatedContactData)
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "UPDATE_ERROR",
                        message: "Failed to update contact",
                        details: error.localizedDescription
                    ))
                }
            }
        }
    }
    
    func handleDeleteContact(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let contactId = args["id"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Contact ID is required", details: nil))
            return
        }
        
        DispatchQueue.global(qos: .userInteractive).async {
            do {
                try self.deleteContact(id: contactId)
                DispatchQueue.main.async {
                    result(nil)
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(
                        code: "DELETE_ERROR",
                        message: "Failed to delete contact",
                        details: error.localizedDescription
                    ))
                }
            }
        }
    }
}